src/tools/ceph-dencoder/CMakeLists.txt

```
@@ -50,7 +50,8 @@ target_link_libraries(denc-mod-osd
   os
   osd
   mon
-  erasure_code)
+  erasure_code
+  global)
```

```suggestion
  erasure_code
  global-static)
```

see 4ea38df656b2aad9398fcb1656c294f5b8366811 for the rationales.

---

I've just tried that, and now `ceph-dencoder` segfaults:

```
# ceph-dencoder 
-h for help
Segmentation fault (core dumped)

# gdb ceph-dencoder 
[...]
Reading symbols from ceph-dencoder...
(No debugging symbols found in ceph-dencoder)
(gdb) run
Starting program: /usr/bin/ceph-dencoder 
Missing separate debuginfos, use: zypper install ceph-common-debuginfo-17.0.0-14884.1.x86_64
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
-h for help

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7de1123 in tcmalloc::ThreadCache::ReleaseToCentralCache(tcmalloc::ThreadCache::FreeList*, unsigned int, int) () from /lib64/libtcmalloc.so.4

```

Note that ceph-dencoder itself links against `global`, not `global-static` (see https://github.com/ceph/ceph/blob/a1bf7f7416f643f8b56fb571f687fa90066cbd86/src/tools/ceph-dencoder/CMakeLists.txt#L99). Could that discrepancy cause this? Should I try changing that to global-static as well, or go back to using `global` for denc-mod-osd?

---

Just for a test, I tried using `global-static` on denc-mod-osd as above, and switched `global` to `global-static` in ceph-dencoder like so:

```diff
--- ceph-17.0.0-14883-gc49b81c7d61.orig/src/tools/ceph-dencoder/CMakeLists.txt
+++ ceph-17.0.0-14883-gc49b81c7d61/src/tools/ceph-dencoder/CMakeLists.txt
@@ -97,7 +97,7 @@ target_compile_definitions(ceph-dencoder
 
 target_link_libraries(ceph-dencoder
   StdFilesystem::filesystem
-  global
+  global-static
   ${DENCODER_EXTRALIBS}
   cls_log_client
   cls_version_client
```

That fixes the segfault, but now I've got another undefined symbol (this time in denc-mod-common):

```
# ceph-dencoder 
failed to dlopen("/usr/lib64/ceph/denc/denc-mod-common.so"): /usr/lib64/ceph/denc/denc-mod-common.so: undefined symbol: _ZTI7Message
-h for help
```

Based on these tests, I think we just need to use `global`, not `global-static`, unless there's some other change(s) necessary in addition to the above.

---

@tserong hi Tim, sorry. i didn't look into the linkage of ceph-dencoder itself. agreed, it it links against `global-static`, we should not link `denc-mod-*` against it again. it'd violate the ODR (one definition rule). but after a second thought, i think we need to drop `global` from ceph-dencoder's linked libraries list. BTW, sorry for the latency. i was in a "power saver mode" last month.

---

#48418 is created to implement this proposal.